pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,172 @@
+        ---------
+        Fixing Max-Product: Convergent Message-Passing Algorithms for MAP LP Relaxations
+        by Amir Globerson and Tommi Jaakkola.
+        Section 6, Page: 5; Beyond pairwise potentials: Generalized MPLP
+        """
+        updated_results = []
+        # The new updates will take place for the intersection_sets of this cluster.
+        for i in sending_cluster.intersection_sets_for_cluster_c:
+
+            # Now we have to construct the crux of the equation which is used to update the messages there.
+
+            # The terms which we need are:
+            # 1. Message not originating from the current sending_cluster but going into the current intersect i
+            message_not_from_cluster = []
+
+            # 2. Summation of messages not originating from the current sending_cluster and going into other
+            #    intersects than i of this cluster only.
+            sum_other_intersects = []
+
+            cluster_list_except_current = [j for j in self.cluster_set if j != sending_cluster]
+            other_intersects = [j for j in sending_cluster.intersection_sets_for_cluster_c if j != i]
+
+            for cluster in cluster_list_except_current:
+                for intersection_set in cluster.intersection_sets_for_cluster_c:
+                    if i == intersection_set:
+                        message_not_from_cluster.append(cluster.message_from_cluster[intersection_set])
+
+                for other_intersect in other_intersects:
+                        for intersection_set in cluster.intersection_sets_for_cluster_c:
+                            if other_intersect == intersection_set:
+                                sum_other_intersects.append(cluster.message_from_cluster[intersection_set])
+
+            # lambda_1 = \lambda_{s}^{-c}
+            lambda_1 = copy.deepcopy(sending_cluster.message_from_cluster[i])
+            lambda_1.values = np.zeros(np.prod(lambda_1.cardinality))
+            # lambda_2 = \sum_{}{\lambda_{s'}^{-c}}
+            lambda_2 = copy.deepcopy(lambda_1)
+            for message in message_not_from_cluster:
+                lambda_1 += message
+            for message in sum_other_intersects:
+                lambda_2 += message
+            # maximize (lambda_2 + sending cluster potential) w.r.t nodes in the present cluster but not in intersect
+            phi = sending_cluster.cluster_potential + lambda_2
+            # phi = max_{x_{c-s}}{lambda_2 + /theta_c}
+            phi.maximize(list(set(sending_cluster.cluster_variables)-i))
+            intersection_length = len(sending_cluster.intersection_sets_for_cluster_c)
+            # \lambda_{c \rightarrow \s} = -(1 - 1 / |S(c)|) lambda_1 + (1 / |S(c)|) phi
+            lambda_c_s = -(1 - 1 / intersection_length) * lambda_1 + (1 / intersection_length) * phi
+            updated_results.append(lambda_c_s)
+
+        for i, j in zip(sending_cluster.intersection_sets_for_cluster_c, range(len(updated_results))):
```

Comment:
@kislayabhi I don't think its a good solution to modify the input variable passed to this method. It would lead to so many bugs which are not trivial to debug. Please think of some more elegant solution if this can't be removed, else remove this.